packages/gatsby/src/query/query-runner.js

```diff
@@ -128,5 +126,12 @@ module.exports = async (graphqlRunner, queryJob: QueryJob) => {
     isPage: queryJob.isPage,
   })
 
+  if (process.env.GATSBY_PAGE_BUILD_ON_DATA_CHANGES && queryJob.isPage) {
+    boundActionCreators.setPageData({
+      id: queryJob.id,
+      result: { resultHash },
```

Comment:
```suggestion
      resultHash,
```
No need to create extra objects if only store hash.

There will be more code changes needed to adjust for changed shape